src/lib/storage/index/partial_hash/partial_hash_index_impl.cpp

```diff
@@ -111,17 +113,27 @@ PartialHashIndexImpl<DataType>::PartialHashIndexImpl(
   insert_entries(chunks_to_index, column_id);
 }
 
+std::mutex insert_entries_mutex;
+
 template <typename DataType>
 size_t PartialHashIndexImpl<DataType>::insert_entries(
     const std::vector<std::pair<ChunkID, std::shared_ptr<Chunk>>>& chunks_to_index, const ColumnID column_id) {
   const auto previous_chunk_count = _indexed_chunk_ids.size();
   for (const auto& chunk : chunks_to_index) {
-    if (_indexed_chunk_ids.contains(chunk.first)) {
+
+    // Prevents multiple threads from indexing the same chunk concurrently.
+    {
+      std::lock_guard<std::mutex> lock{insert_entries_mutex};
```

Comment:
Instead of locking for each chunk, I think it is sufficient to lock only when the index does not contain index entries for a given chunk, e.g.,:

```
if contains:
  continue

// now, we have to check again but locked this time
lock_guard
// check again since another thread could have added it already
if !contains:
  insert
```

Reply:
Like this?
```
    if (_indexed_chunk_ids.contains(chunk.first)) {
      // Index already contains entries for the given chunk.
      continue;
    }

    // Prevents multiple threads from indexing the same chunk concurrently.
    {
      std::lock_guard<std::mutex> lock{insert_entries_mutex};
      
      if (_indexed_chunk_ids.contains(chunk.first)) {
      // Another thread could have added the same chunk in the meantime.
      continue;
      }

      _indexed_chunk_ids.insert(chunk.first);
    }
```
I don't really like the redundant if, but if it leads to a better performance it would be fine of course.

Reply:
`auto lock = std::lock_guard<std::mutex>{insert_entries_mutex};`? :b